lib/open_telemetry/honeycomb/event.ex

```diff
@@ -0,0 +1,194 @@
+
+  @typedoc """
+  Honeycomb event suitable for POSTing to their batch API.
+
+  * `time`: ms since epoch; [MUST] be in ISO 8601 format, e.g. `"2019-05-17T09:55:12.622658Z"`
+  * `data`: `t:event_data/0` after flattening.
+
+  [MUST]: https://tools.ietf.org/html/rfc2119#section-1
+  """
+  @type t :: %__MODULE__{
+          time: String.t(),
+          data: event_data()
+        }
+
+  @doc """
+  The current UTC time in ISO 8601 format, e.g. `"2019-05-17T09:55:12.622658Z"`
+
+  Useful when creating events manually.
+  """
+  @spec now() :: String.t()
+  def now do
+    DateTime.utc_now() |> DateTime.to_iso8601()
+  end
+
+  @doc """
+  Convert one OpenTelemetry span to an event suitable for POSTing to the
+  [Honeycomb Events API][HCevents].
+
+  See [attribute limitations](#module-opentelemetry-honeycomb) for important detail on span a
+  attribute names and values.
+
+  [HCevents]: https://docs.honeycomb.io/api/events/
+  [HCstruct]: https://github.com/honeycombio/opentelemetry-exporter-go/blob/66047b9/honeycomb/honeycomb.go#L325
+  [HCdefs]: https://docs.honeycomb.io/working-with-your-data/managing-your-data/definitions/#tracing
+  """
+  @spec from_ot_span(
+          :opentelemetry.span(),
+          resource_attributes :: OpenTelemetry.attributes(),
+          attribute_map :: AttributeMap.t()
+        ) :: [t()]
+  def from_ot_span(ot_span, resource_attributes, attribute_map) do
+    span = Span.from(ot_span)
+
+    data =
+      span
+      |> Map.get(:attributes)
+      |> merge(resource_attributes)
+      |> merge(extracted_attributes(span, attribute_map))
+      |> Enum.into(%{})
+      |> Cleaner.clean()
+
+    time =
+      span.start_time
+      |> :opentelemetry.convert_timestamp(:microsecond)
+      |> DateTime.from_unix!(:microsecond)
+      |> DateTime.to_iso8601()
+
+    [%__MODULE__{time: time, data: data}]
+  end
+
+  @spec merge(OpenTelemetry.attributes(), OpenTelemetry.attributes()) ::
+          OpenTelemetry.attributes()
+  defp merge(att1, att2), do: :lists.ukeymerge(1, att1, att2)
+
+  # span attribute extractors
+  @spec extracted_attributes(Span.t(), AttributeMap.t()) :: OpenTelemetry.attributes()
+  defp extracted_attributes(%Span{} = span, attribute_map) do
+    attribute_mapper = get_attribute_mapper(attribute_map)
+
+    [
+      duration_ms: ms(span.end_time) - ms(span.start_time),
+      name: span.name,
+      parent_span_id: hexify_span_id(span.parent_span_id),
+      span_id: hexify_span_id(span.span_id),
+      trace_id: hexify_trace_id(span.trace_id)
+    ]
+    |> Enum.map(attribute_mapper)
+    |> Enum.filter(&has_binary_key?/1)
+  end
+
+  @spec get_attribute_mapper(attribute_map :: AttributeMap.t()) ::
+          ({atom(), OpenTelemetry.attribute_value()} ->
+             {OpenTelemetry.attribute_key(), OpenTelemetry.attribute_value()})
+  defp get_attribute_mapper(map) do
+    fn {k, v} -> {Map.get(map, k), v} end
+  end
```

Comment:
Had to go back to `AttributeMap` to understand why this thing exists and looks so strange. My first reaction was "why would those keys be in the span attributes at all?" Instead what you've done is defined naming translations for keys that are in the spans into Honeycomb's conventions.

Maybe naming the function (and AttributeMap module) differently would help, or even inlining it into the function that calls this one.

Reply:
Six lines of overhead to move a closure is a OTT, yeah. Would it make sense to move it to `AttributeMap`. If that makes sense, would we also move this `extracted_attributes/2` to `OpenTelemetry.Records.Span`? Feels right but it screws up the move I had in mind of moving `OpenTelemetry.Records.{Event,Span}` to the API package to save time for other Elixir programmers who need to pull those records apart.